R/cube-comparisons.R

```diff
@@ -0,0 +1,299 @@
+    }
+    out <- vapply(names, function(cur) {
+            compareDimsPairwise(cube, dim=dim, baseline=cur, value=value)
+        },
+        numeric(length(names))
+    )
+    return(simplify2array(out))
+}
+
+#' Matrix of Chi-Squared Statistics for all rows or columns
+#' 
+#' Use the alternative Wishart method of forming the matrix of column- or row-wise
+#' comparison Chi-squared test statistics for a categorical-by-categorical 
+#' contingency table.
+#'
+#' @param cube a CrunchCube of counts: a crosstab of two categorical variables
+#' @param dim dimension along which to compare proportions (default `cols` or `rows`)
+#'
+#' @return A symmetric square matrix of Chi-squared statistics for a crosstab
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' # TODO: include Hirotsu occupation-illness example cube and output
+#' illness_occupation <- crunch_example_data(hirotsu)
+#' # Chi-squared statistic for joint null hypothesis that each
+#' # column is equal to each other column.
+#' 
+#' wishartMatrix(illness_occupation)
+#' }
+wishartMatrix <- function(cube, dim=c("cols","rows")){
+    dim <- match.arg(dim)
+    a <- as.array(cube)
+    if(dim == "rows") {
+        # This would be simpler if we could locally transpose the CrunchCube,
+        # but instead these results will have transposed shapes for rows.
+        a <- t(a)
+        props <- t(prop.table(cube, 1))
+        margins <- margin.table(cube, 1)
+        weights <- t(prop.table(margin.table(cube, 2)))
+    } else {
+        props <- prop.table(cube, 2)
+        margins <- margin.table(cube, 2)
+        weights <- prop.table(margin.table(cube, 1))
+    }
+    names <- dimnames(a)[[2]]
+    
+    
+    I <- ncol(a)
+    X2 <- matrix(0.0, nrow = I, ncol = I)
+    for (i in 2:I){
+            for (j in 1:(i-1)) {
+                X2[j,i] <- X2[i,j] <- sum( (props[,i] - props[,j])^2 / weights ) /
+            ( 1 / margins[i] + 1 / margins[j] )
+            }
+    }
+    if(dim == "rows") {
+        X2 <- t(X2)
+    }
+    dimnames(X2) <- list(names, names)
+    return(X2)
+}
+
+
+#' P-values for pairwise comparison on a dimension of a categorical crosstab
+#'
+#' @param cube a CrunchCube
+#' @param dim dimension which jointly forms the null hypothesis that each
+#' vector is equal to all of the others.
+#'
+#' @return A symmetric square matrix of the size of `dim`. Typical element
+#' \eqn{i,j} is the P-value associated with the hypothesis that proportions in
+#' row or column \eqn{i} are equal to those in row or column \eqn{j}.
+#' 
+#' @importFrom rootWishart singleWishart
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' # TODO: include Hirotsu occupation-illness example cube and output
+#' illness_occupation <- crunch_example_data(hirotsu)
+#' # P-value for joint null hypothesis that each
+#' # column is equal to each other column.
+#' 
+#' wishartPvalues(illness_occupation)
+#' }
+wishartPvalues <- function(cube, dim=c("cols", "rows")){
+    X2 <- wishartMatrix(cube, dim)
```

Comment:
If rootWishart is `Suggests`, don't import the functions, call them with `rootWishart::fun()`, and also check that `rootWishart` is installed (see https://github.com/Crunch-io/rcrunch/blob/master/R/misc.R#L288)